analyzers/src/SonarAnalyzer.CSharp/Rules/TooManyLoggingCalls.cs

```diff
@@ -0,0 +1,217 @@
+            logCallCollector.Visit(node);
+            foreach (var loggingCategory in logCallCollector.LoggingInvocations)
+            {
+                var treshold = Treshold(loggingCategory.Key);
+                var invocations = loggingCategory.Value;
+                if (invocations.Count > treshold)
+                {
+                    var primaryLocation = invocations[0].GetLocation();
+                    var secondaryLocations = invocations.Skip(1).Select(x => x.GetLocation());
+                    context.ReportIssue(Diagnostic.Create(Rule, primaryLocation, secondaryLocations, loggingCategory.Key, invocations.Count, treshold));
+                }
+            }
+        }
+    }
+
+    private int Treshold(string category)
+    {
+        var threshold = category switch
+        {
+            CategoryNames.Debug => DebugThreshold,
+            CategoryNames.Error => ErrorThreshold,
+            CategoryNames.Warning => WarningThreshold,
+            _ => InformationThreshold
+        };
+        return Math.Max(threshold, 0);
+    }
+
+    private static bool IsBlockNodeSupported(SyntaxNode node) =>
+        node.Kind() == SyntaxKind.Block
+        || (node.Kind() == SyntaxKind.CompilationUnit && node.ChildNodes().OfType<GlobalStatementSyntax>().Any());  // for top-level statements
+
+    private sealed class LoggingCallCollector : SafeCSharpSyntaxWalker
+    {
+        private readonly SemanticModel semanticModel;
+        private SyntaxNode currentBlock;
+
+        public Dictionary<string, List<InvocationExpressionSyntax>> LoggingInvocations { get; } = [];
+
+        public LoggingCallCollector(SemanticModel semanticModel, SyntaxNode currentBlock)
+        {
+            this.semanticModel = semanticModel;
+            this.currentBlock = currentBlock;
+        }
+
+        public override void VisitBlock(BlockSyntax node)
+        {
+            if (currentBlock == node) // Don't visit nested blocks, that will be done by additional LoggingCallCollector instances.
+            {
+                currentBlock = node;
+                base.VisitBlock(node);
+            }
+        }
+
+        public override void VisitInvocationExpression(InvocationExpressionSyntax node)
+        {
+            if (IsLoggerMethod(node.GetName())
+                && semanticModel.GetSymbolInfo(node).Symbol is IMethodSymbol methodSymbol
+                && methodSymbol.ContainingType.DerivesFromAny(LoggerTypes)
+                && LoggingCategoryName(node, methodSymbol) is { } loggingCategory)
+            {
+                if (LoggingInvocations.TryGetValue(loggingCategory, out var invocationList))
+                {
+                    invocationList.Add(node);
+                }
+                else
+                {
+                    LoggingInvocations.Add(loggingCategory, [node]);
+                }
+            }
+        }
+
+        public override void VisitIfStatement(IfStatementSyntax node)
```

Comment:
Same question that I asked on `Lambda` overrides.
`If` generates their own code block, which you will short-circuit in your `VisitBlock` logic.

Reply:
Just like the lamda, an if statement doesn't always use a block (although it's good practice).
[See this potential FP](https://github.com/SonarSource/sonar-dotnet/blob/18e46259cfe3eb14695507014b26fc2674669020/analyzers/tests/SonarAnalyzer.Test/TestCases/TooManyLoggingCalls.cs#L192).